runtime/service.go

```diff
@@ -436,10 +345,266 @@ func (s *service) nextVSockPort() uint32 {
 	return port
 }
 
+func (s *service) waitVMReady() error {
+	select {
+	case <-s.vmReady:
+		return nil
+	case <-time.After(firecrackerStartTimeout):
+		return status.Error(codes.DeadlineExceeded, "timed out waiting for VM start")
+	}
+}
+
+// CreateVM will attempt to create the VM as specified in the provided request, but only on the first request
+// received. Any subsequent requests will be ignored and get an AlreadyExists error response.
+func (s *service) CreateVM(requestCtx context.Context, request *proto.CreateVMRequest) (*empty.Empty, error) {
+	defer logPanicAndDie(s.logger)
+
+	errCh := make(chan error, 1)
+	s.vmStartOnce.Do(func() {
+		errCh <- s.createVM(requestCtx, request)
+		close(errCh)
+	})
+
+	select {
+	case err := <-errCh:
+		if err != nil {
+			// if we fail to create the VM, we have no point in existing anymore, so shutdown
+			s.shimCancel()
+			err = errors.Wrap(err, "failed to create VM")
+			s.logger.WithError(err).Error()
+			return nil, err
+		}
+
+		// creating the VM succeeded, setup monitors and publish events to celebrate
+
+		if request.ExitAfterAllTasksGone {
+			go s.monitorTaskExit()
+		}
+
+		go func() {
```

Comment:
Can the event stuff be moved out into a separate function?

Reply:
Each individual function (the `Attach` goroutine, the `VMStart` publish and the `VMStop` publish) can be separately factored out into their own function, but it wouldn't really be possible or make sense to put all or a combination of them in one single function.

I can certainly do that, but it's one of those situations where we would be factoring out 3 line functions that only get called exactly once in the whole program. It doesn't provide any DRY benefits and doesn't IMO help readability that much because they are so short in the first place. I feel like all it does is make it harder to know what's actually happening since it forces you to jump around to random functions across the codebase instead of just seeing them directly in the only place they are actually used.

This is another case of splitting hairs though, if you or anyone else feels strongly I have no problem factoring these out.

Reply:
What about something like this?

```go
func (s *service) CreateVMEvents(vmid string) (createFunc, stopFunc func() error) {
	go func() {
		if bridgeErr := eventbridge.Attach(s.shimCtx, s.eventBridgeClient, s.eventExchange); bridgeErr != nil {
			s.logger.WithError(err).Error("error while forwarding events from VM to agent")
		}
	}()
	createFunc = func() error {
		return s.eventExchange.Publish(s.shimCtx, StartEventName, &proto.VMStart{VMID: vmid})
	}
	stopFunc = func() error {
		return s.eventExchange.Publish(s.shimCtx, StopEventName, &proto.VMStop{VMID: vmid})
	}
	return
}
```

I don't really like mixing the nuts and bolts of how an event is created (protobufs) with the logic of actually starting the VM.

Reply:
We can consider runc shim's approach, that makes events publishing a lot more cleaner ([1](https://github.com/containerd/containerd/blob/master/runtime/v2/runc/v2/service.go#L295), [2](https://github.com/containerd/containerd/blob/master/runtime/v2/runc/v2/service.go#L690))
But I'm not sure that it worth doing that for a just few events.

Reply:
I get the point about not mixing in the proto stuff with everything else, so I just added two methods `publishVMStart` and `publishVMStop`. The names are obvious enough that I think it's fine to move the code elsewhere even though they are only called once each.

For the rest, I ended up taking a slightly different approach that I believe addresses the concerns:
1. I moved the `eventbridge.Attach` call to a common method with `eventbridge.Republish`. This makes logical sense in that now all of the republishing/forwarding we do is setup in one place.
2. I created a method similar to the existing `monitorTaskExit` called `monitorVMExit` that does that `Wait` call and publishes the Stop event once that returns.

Let me know if that addresses the concerns.